photonpump/connection.py

```diff
@@ -759,58 +834,213 @@ def close(self):
             consumer_strategy=consumer_strategy
         )
 
-        future = await self.protocol.enqueue_conversation(cmd)
+        future = await self.dispatcher.start_conversation(cmd)
 
         return await future
 
-    async def connect_subscription(self, subscription: str, stream: str):
+    async def connect_subscription(
+            self,
+            subscription: str,
+            stream: str,
+            conversation_id: Optional[uuid.UUID] = None
+    ):
         cmd = convo.ConnectPersistentSubscription(
-            subscription, stream, credentials=self.credential
+            subscription,
+            stream,
+            credentials=self.credential,
+            conversation_id=conversation_id
         )
-        future = await self.protocol.enqueue_conversation(cmd, True)
+        future = await self.dispatcher.start_conversation(cmd)
 
         return await future
 
-    async def ack(self, subscription, message_id, correlation_id=None):
-        cmd = msg.AcknowledgeMessages(
-            subscription, [message_id],
+    async def ack(self, subscription, message_ids, correlation_id=None):
+        cmd = msg.PersistentSubscriptionAckEvents(
+            subscription,
+            message_ids,
             correlation_id,
             credentials=self.credential,
             loop=self.loop
         )
-        await self.protocol.enqueue_conversation(cmd)
+        await self.dispatcher.start_conversation(cmd)
+
+    async def send_heartbeats(self):
+        heartbeat_id = uuid.uuid4()
+
+        while True:
+            logging.debug("Sending heartbeat %s to server", heartbeat_id)
+            hb = convo.Heartbeat(
+                heartbeat_id, direction=convo.Heartbeat.OUTBOUND
+            )
+            fut = await self.dispatcher.start_conversation(hb)
 
+            try:
+                await asyncio.wait_for(fut, 10)
```

Comment:
so `self.dispatcher.start_conversation(hb)` is a coroutine that returns a `Future`.

What is this `future` about?

Reply:
Okay, so when we create a new conversation, we give it a Future for its result. This will eventually have set_result or set_exception called on it to return the response from the server.

start_conversation is a co-routine that returns that future. It's used by the client like this:

```
async def ping():
    # create a new conversation
    conversation = Ping()

    # start it, by giving it to the dispatcher for sending
    fut = await dispatcher.start_conversation(conversation)

    # wait for the returned exception to complete
    return await fut
```

When we finish #51 that'll change slightly to this:

```
async def ping():
    # create a new conversation
    conversation = Ping()

    # start it, by giving it to the dispatcher for sending
    await dispatcher.start_conversation(conversation)

    # wait for the conversation to complete
    return await conversation.result
```

Reply:
Thanks!